recipes/MEDIA/SLU/CTC/train_hf_wav2vec.py

```diff
@@ -0,0 +1,390 @@
+> python train_with_wav2vec.py hparams/train_with_wav2vec.yaml
+
+With the default hyperparameters, the system employs a VanillaNN encoder.
+
+The neural network is trained on greedy CTC.
+
+The experiment file is flexible enough to support a large variety of
+different systems. By properly changing the parameter files, you can try
+different encoders, decoders, training tasks (Media , PortMedia),
+and many other possible variations.
+
+Authors
+ * Gaelle Laperriere 2023
+"""
+
+import sys
+import torch
+import logging
+import speechbrain as sb
+from hyperpyyaml import load_hyperpyyaml
+from speechbrain.dataio.batch import PaddedBatch
+from speechbrain.utils.distributed import run_on_main
+from media_prepare import prepare_media
+
+logger = logging.getLogger(__name__)
+
+
+# Define training procedure.
+class SLU(sb.core.Brain):
+    def compute_forward(self, batch, stage):
+        """Forward computations from waveform to output probabilities."""
+
+        # Get data.
+        batch = batch.to(self.device)
+        wavs, wav_lens = batch.sig
+        wavs, wav_lens = wavs.to(self.device), wav_lens.to(self.device)
+
+        # Forward pass.
+        feats = self.modules.wav2vec2(wavs, wav_lens)
+
+        x = self.modules.enc(feats)
+
+        # Output layer for seq2seq log-probabilities.
+        logits = self.modules.output_lin(x)
+        p_ctc = self.hparams.log_softmax(logits)
+
+        return p_ctc, wav_lens
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the loss (CTC) given predictions and targets."""
+
+        # Get data.
+        batch = batch.to(self.device)
+        chars, char_lens = batch.char_encoded
+        ids = batch.id
+
+        # Get predictions & loss.
+        p_ctc, wav_lens = predictions
+        loss = self.hparams.ctc_cost(p_ctc, chars, wav_lens, char_lens)
+
+        # Get metrics.
+        if stage != sb.Stage.TRAIN:
+            # Generate sequences with CTC greedy decoder.
+            sequence = sb.decoders.ctc_greedy_decode(
+                p_ctc, wav_lens, self.hparams.blank_index
+            )
+            # Update metrics.
+            self.cer_metric.append(
+                ids=ids,
+                predict=sequence,
+                target=chars,
+                target_len=char_lens,
+                ind2lab=self.label_encoder.decode_ndim,
+            )
+            self.coer_metric.append(
+                ids=ids,
+                predict=sequence,
+                target=chars,
+                target_len=char_lens,
+                ind2lab=self.label_encoder.decode_ndim,
+            )
+            self.cver_metric.append(
+                ids=ids,
+                predict=sequence,
+                target=chars,
+                target_len=char_lens,
+                ind2lab=self.label_encoder.decode_ndim,
+            )
+            self.ctc_metric.append(ids, p_ctc, chars, wav_lens, char_lens)
+
+        return loss
+
+    def fit_batch(self, batch):
```

Comment:
same as mentioned earlier, might be neat if you could add also the support for torch.amp